File: TicTacToe/C/TicTacToe.c
#include <stdio.h>		/*	printf			*/
#include <stdlib.h>		/*	rand, malloc	*/
#include <time.h>		/*	time			*/

// use: gcc -Wall -pedantic -o TicTac TicTacToe.c

/* Type Definitions */
typedef enum {false, true} bool;
typedef enum {X_ = -1, empty = 0, O_ = 1} TickTack;
typedef struct {
	int horizontal;
	int vertical;
	int diagonal;
	int antidiag;
} checkCount;

/* Constant global variables	*/
	/* TickTacks */
	char TickTacks[3] = {'x', ' ', 'o'};
	TickTack *gameBoard;

// Function proptypes declarations
void menu(TickTack *piece, bool *AI, bool *rand);
void initializeGameBoard();
int runGame_TicTacToe(TickTack piece, bool AI, bool rand);
int playerTurn(TickTack piece);
int AITurn(TickTack piece, bool rand);
double minimax(int startCoords, TickTack piece, bool isMaximizing, bool rand, double alpha, double beta);
int checkGameState_full();
void displayBoard();

// Standard Function proptypes declarations
void toLower(char* str);
int strcmp(const char* str1, const char* str2);
int strti(const char* str, int *error);
void flush();
double doubleRand(double min, double max);

/* Macros */
#define TIE 3
#define POS_INF 1.0 / 0.0
#define NEG_INF -1.0 / 0.0
#define swap(x, y) do {typeof(x) temp = x; x = y; y = temp;} while(0);

/*
 * function_identifier: The program's main function.
 * parameters: 			N/A
 * return value:		The exit status of the program.
*/
int main(int argc, char *argv[]) {
	// Initializes variables
	int piece;
	bool AI, rand;
	srand(time(NULL));
	// Call the starting user GUI
	menu(&piece, &AI, &rand);

	// Plays the game and displays message appropriate to the game's ending status
	switch (runGame_TicTacToe(piece, AI, rand)) {
		case X_:	printf("X wins!\n"); return 0;
		case O_:	printf("O wins!\n"); return 0;
		case TIE:	printf("It's a Tie!\n"); return 0;
		default:	printf("Some unexpected error occured.\n"); return 1;
	}
}

/*
 * function_identifier: The starting GUI for the client
 * parameters: 			The address of a TickTack enum, and the address of an integer
 * return value:		N/A
*//*
 * function_identifier: The starting GUI for the client
 * parameters: 			The address of a TickTack enum, and the address of an integer
 * return value:		N/A
*/
void menu(TickTack *piece, bool *AI, bool *rand) {
	// Initializes variable
	char str[10];

	// Displays user prompt
	printf("Do you want to play with Ai or another user?\n1.) AI\n2.) User\n");

	// Second user prompt loop
	promptJump1:
		printf("Please enter your selection.\n");

		scanf("%s", str); 	// Gets a single character from the I/O buffer. Ignores newlines
		toLower(str);	// Converts string to lowercase

		// If the user did give a vaild character, set the piece to the appropriate value
		if (!(strcmp(str, "1") && strcmp(str, "ai"))) *AI = true;
		else if (!(strcmp(str, "2") && strcmp(str, "u") && strcmp(str, "user"))) *AI = false;
		else {
			// If the user did not give a vaild character, display error message and jump back
			printf("Invaild selection. Please try again.\n"); // 
			flush(); // Clears the I/O buffer
			goto promptJump1;
		}

	// Returns if no AI
	if (!*AI) {*piece = X_; *rand = false; return;}

	// Displays user prompt
	printf("Which piece will you play?\n1.) X\n2.) O\n");

	// First user prompt loop
	promptJump2:
		printf("Please enter your selection.\n");

		scanf("%s", str);	// Gets a single character from the I/O buffer. Ignores newlines
		toLower(str);	// Converts string to lowercase

		// If the user did give a vaild character, set the piece to the appropriate value
		if (!(strcmp(str, "1") && strcmp(str, "x"))) *piece = X_;
		else if (!(strcmp(str, "2") && strcmp(str, "o"))) *piece = O_;
		else {
			// If the user did not give a vaild character, display error message and jump back
			printf("Invaild selection. Please try again.\n"); // 
			flush(); // Clears the I/O buffer
			goto promptJump2;
		}

	// Displays user prompt
	printf("What difficulty do you wish to play at?\n1.) Easy\n2.) Hard\n");

	// Third user prompt loop
	promptJump3:
		printf("Please enter your selection.\n");

		scanf("%s", str);	// Gets a single character from the I/O buffer. Ignores newlines
		toLower(str);	// Converts string to lowercase

		// If the user did give a vaild character, set the piece to the appropriate value
		if (!(strcmp(str, "1") && strcmp(str, "e") && strcmp(str, "easy"))) *rand = true;
		else if (!(strcmp(str, "2") && strcmp(str, "h") && strcmp(str, "hard"))) *rand = false;
		else {
			// If the user did not give a vaild character, display error message and jump back
			printf("Invaild selection. Please try again.\n"); // 
			flush(); // Clears the I/O buffer
			goto promptJump3;
		}
}

/*
 * function_identifier: The starting GUI for the client
 * parameters: 			The address of a TickTack enum, and the address of an integer
 * return value:		The ending status of the game
*/
int runGame_TicTacToe(TickTack ticktack, bool AI, bool rand) {
	// Initializes variables and game board
	int result = empty, piece = X_;
	initializeGameBoard();

	// Displays the empty game board
	displayBoard();
	while(1) {
		// Plays each turn
			// A user turn
		if (piece == ticktack || !AI) {printf("\nPlayer %d's turn!\n", ((piece != ticktack)+1)); result = playerTurn((TickTack)piece);}
			// An AI turn
		else {printf("\nComputer's turn!\n"); result = AITurn((TickTack)piece, rand);}
		// Displays the game board after a turn
		displayBoard();
		// Ends the game when a conclution has occured
		switch(result) {
			case empty: break;
			case X_:
			case O_:
			case TIE:
				free(gameBoard);
				return result;
		}
		// Inverts the piece
		piece = -piece;
	}
	// If this instruction ever executes, an error occured
	return -2;
}

/*
 * function_identifier: Initializes the game board's values.
 * parameters: 			N/A
 * return value:		N/A
*/
void initializeGameBoard() {
	// Allocates memory for the game board
	gameBoard = (TickTack*)malloc(9*sizeof(TickTack));
	// Initializes allocated memory
	for (int i=0; i<9; i++) {gameBoard[i]=empty;}
}

/*
 * function_identifier: Plays a turn according to user input
 * parameters: 			The player's current piece
 * return value:		The status of the board after the player finished their turn
*/
int playerTurn(TickTack piece) {
	// Initializes variables and game board
	int X, Y, coords;

	// The prompt loop
	promptJump4:
		printf("Enter X and Y coords: ");
		scanf("%d%d", &X, &Y); // Scans the I/O buffer for integers
		if (X > 3 || X < 1) {
			// If an invalid X coordinate is given, display error and jump back to loop
			printf("Error - X coordinate must between 1-3.\n\n");
			flush(); // Clears the I/O buffer
			goto promptJump4;
		}
		if (Y > 3 || Y < 1) {
			// If an invalid Y coordinate is given, display error and jump back to loop
			printf("Error - Y coordinate must between 1-3.\n\n");
			flush(); // Clears the I/O buffer
			goto promptJump4;
		}
		if (gameBoard[(coords = X + Y*3 - 4 )] != empty) {
			// If an invalid coordinate is given, display error and jump back to loop
			printf("Error - coordinate position (%d, %d) is nonempty.\n\n", X, Y);
			flush(); // Clears the I/O buffer
			goto promptJump4;
		}

	gameBoard[coords] = piece; // Sets the coordinate
	return checkGameState_full(); // Return the gameBoard's status
}

/*
 * function_identifier: Plays a turn according to minMax algorithm
 * parameters: 			The AI's current piece
 * return value:		The status of the board after AI finished their turn
*/
int AITurn(TickTack piece, bool rand) {
	// Initializes variables and game board
	int bestMove = -2;
	double bestScore = -2, score;
	// If the given piece is empty, then just return TIE result
	if (piece == empty) {return TIE;}

	// Simulates placing a piece on all given board positions
	for (int coords=8; coords>=0; --coords) {
		if (gameBoard[coords] == empty) {
			score = minimax(coords, piece, false, rand, NEG_INF, POS_INF);
			if (score > bestScore) {
				bestScore = score;
				bestMove = coords;
			}
		}
	}

	// If no bestMove was found possible, error
	if (bestMove == -2) printf("ERROR\n");
	// Plays out the given best move
	gameBoard[bestMove] = piece;
	// Returns the current status of the game board
	return checkGameState_full();
}

/*
 * function_identifier: The minMax algorithm of a TickTacToe board
 * parameters: 			The AI's current piece
 * return value:		The vaild of the current board the algorithm is examining
*/
double minimax(int startCoords, TickTack piece, bool isMaximizing, bool rand, double alpha, double beta) {
	// Plays the given piece
	gameBoard[startCoords] = (isMaximizing) ? (TickTack)(-piece) : piece;
	// Checks the board's current state
	int result = checkGameState_full();
	// Ends this simulation branch if an end result is found
	if (result != empty) {
		gameBoard[startCoords] = empty;
		if (result==TIE) {return 0;}
		if (result==piece) {return 1;}
		return -1;
	}

	// Gets randomised value for lower difficulty
	double random = rand ? doubleRand(0, 0.99) : 0;

	// Initializes variables
	int coords;
	double bestScore, score;
	//	Acting player (max)
	if (isMaximizing) {
		bestScore = -2;
		// Simulates placing a piece on all given board positions for the acting player
		for (coords=9; coords>=0; --coords) {
			if (gameBoard[coords] == empty) {
				score = minimax(coords, piece, false, rand, alpha, beta) + random;
				// Alpha-Beta Pruning
				if (score > bestScore) {bestScore = score;}
				if (score > alpha) {alpha = score;}
				if (beta <= alpha) {break;}
			}
		}
	}
	//	Enemy Player (min)
	else {
		bestScore = 2;
		// Simulates placing a piece on all given board positions for the enemy player
		for (coords=9; coords>=0; --coords) {
			if (gameBoard[coords] == empty) {
				score = minimax(coords, piece, true, rand, alpha, beta) - random;
				// Alpha-Beta Pruning
				if (score < bestScore) {bestScore = score;}
				if (score < beta) {beta = score;}
				if (beta <= alpha) {break;}
			}
		}
	}

	// Ends this branch's simulation
	gameBoard[startCoords] = empty;
	return bestScore;
}

/*
 * function_identifier: Checks the state of the current game board
 * parameters: 			N/A
 * return value:		The current game board's state
*/
int checkGameState_full() {
	// Initializes variables
	int X, Y, temp, check, isTie = 1;

	// Row Check
	for (Y = 0; Y< 3; Y++) {
		temp = Y*3;
		// Checks all rows for three pieces of the same
		if (gameBoard[temp] != empty) {
			check = -gameBoard[temp];
			// If three pieces of the same
				// If so, return that end result
			if (gameBoard[temp] == gameBoard[temp + 1] && gameBoard[temp + 1] == gameBoard[temp + 2]) {return gameBoard[temp];}
			// Checks if there is a row where a piece could still win in
			if (gameBoard[temp + 1] != check && gameBoard[temp + 2] != check) {isTie = 0;}
			continue;
		}
		// Checks if there is a row where a piece could still win in
		if (gameBoard[temp + 1] == gameBoard[temp + 2]) {isTie = 0;}
	}
	// Column Check
	for (X = 0; X < 3; X++) {
		temp = X;
		// Checks all columns for three pieces of the same
		if (gameBoard[temp] != empty) {
			check = -gameBoard[temp];
			// If three pieces of the same
				// If so, return that end result
			if (gameBoard[temp] == gameBoard[temp + 3] && gameBoard[temp + 3] == gameBoard[temp + 6]) {return gameBoard[temp];}
			// Checks if there is a column where a piece could still win in
			if (gameBoard[temp + 3] != check && gameBoard[temp + 6] != check) {isTie = 0;}
			continue;
		}
		// Checks if there is a column where a piece could still win in
		if (gameBoard[temp + 3] == gameBoard[temp + 6]) {isTie = 0;}
	}
	// Diagonal Checks
	for (X = -1; X < 2; X += 2) {
		// Checks all diagonals for three pieces of the same
		if (gameBoard[4] != empty) {
			check = -gameBoard[4];
			// If three pieces of the same
				// If so, return that end result
			if (gameBoard[4] == gameBoard[1 + X] && gameBoard[4] == gameBoard[7 - X]) {return gameBoard[4];}
			// Checks if there is a diagonal where a piece could still win in
			if (gameBoard[1 + X] != check && gameBoard[7 - X] != check) {isTie = 0;}
			continue;
		}
		// Checks if there is a diagonal where a piece could still win in
		if (gameBoard[1 + X] == gameBoard[7 - X]) {isTie = 0;}
	}

	// If there are no rows, columns, or diagonal where a piece could win, return TIE result
	if (isTie) return TIE;
	// Else, return empty result
	return empty;
}

/*
 * function_identifier: Displays the current gameBoard
 * parameters: 			N/A
 * return value:		N/A
*/
void displayBoard() {
	// Prints game board
	printf(" ===========\n");

	for (int Y = 0; Y < 3; Y++) {
		printf("| ");
		for (int X = 0; X < 3; X++) {
			printf("%c | ", TickTacks[1 + gameBoard[ X + Y * 3 ]]);
		}
		printf("\n ===========\n");
	}
}

/*
 * function_identifier: Converts a string into lowercase
 * parameters: 			A string (or a const char array)
 * return value:		N/A
*/
void toLower(char* str) {
	// Initializes variables
	int length = sizeof(str);
	// Cycles through the given strings and converts all applicable characters to lower
	for (int i = 0; i < length; i++) {
		if (90 >= str[i] && str[i] >= 65) str[i] = str[i] + 32;
	}
}

/*
 * function_identifier: Compairs two strings
 * parameters: 			Two strings (or two const char arrays)
 * return value:		returns 0 if (str1 == str2), 1 if (str1 > str2), or -1 if (str1 < str2)
*/
int strcmp(const char* str1, const char* str2) {
	// Initializes variables
	int val=0, i=-1;
	// Cycles through the given strings
	do {
		i++;
		// Compairs character number i in both strings
		val = (str1[i] || str2[i]) * (((str1[i] > str2[i]) || !str2[i]) - ((str2[i] > str1[i]) || !str1[i]));
	} while (!val && str1[i] && str2[i]);	// Stops cycling if one of the strings end or an inequality was detected
	return val;		// Return comparison results
}

/*
 * function_identifier: Converts a string into a signed or unisigned integer
 * parameters: 			Two strings (or two const char arrays)
 * return value:		returns 0 if (str1 == str2), 1 if (str1 > str2), or -1 if (str1 < str2)
*/
int strti(const char* str, int *error) {
	// Initializes variables
	int i, sign=1, signedMax = 0x7FFFFFFF;
	unsigned int val=0, unsignedMax = 0xFFFFFFFF;

	for (i=0; (str[i] == ' '); i++ );		// Ignores all leading spaces
	if (!str[i]) {*error = 2; return 0;}	// Returns with error if string has no non-space chars
	if (str[i] == '-') {sign=-1; i++;}		// Sets "sign" to negative if '-' character is encountered before any numerical chars

	for (;str[i];i++) { // Ends if encounters a NULL character
		if (((str[i] > '9') || (str[i] < '0'))) {*error = i; return val;}		// returns calculated result and error if non-numerical char encountered
		// returns signedMax and error if integer overflow will occur
		if (sign < 0) {if ( val > ((signedMax-(str[i]-'0'))/10)) {*error = 3; return signedMax;}}
		// returns unsignedMax and error if unsigned integer overflow will occur
		else if (val > ((unsignedMax-(str[i]-'0'))/10)) {*error = 3; return unsignedMax;}
		
		// value = (value*10) + next_number_in_string_sequence
		val *= 10;
		val += (str[i]-'0');
	}

	// Returns with no error and converted value
	*error = 0;
	return val*sign;
}

/*
 * function_identifier: Clears the I/O terminal buffer
 * parameters: 			N/A
 * return value:		N/A
*/
void flush (){
	int c;
	// Takes all characters out of I/O buffer until end of line found
	while ((c = getchar()) != '\n' && c != EOF);
}

/*
 * function_identifier: Clears the I/O terminal buffer
 * parameters: 			N/A
 * return value:		N/A
*/
double doubleRand(double min, double max) {
	// Swaps parameters if min is greater than max
	if (min > max) swap(min, max);
	// Return random double between range
    return (min + (rand() / (RAND_MAX / (max - min))));
}

File: TicTacToe/Python/TicTacToe.py
def displayBoard():
	"""
	Uses ASCII text to display the current gameBoard
	"""
	# Checks if there is a board to display
	if len(gameBoard) != 0:
		print(' ===', '====' * (len(gameBoard) - 1), sep='') 		# Top border
		for row in gameBoard:
			print('| ', end='')
			for col in row:
				print(sprites[col], end=' | ')						# Displays the pieces
			print('')
			print(' ===', '====' * (len(gameBoard) - 1), sep='')	# Bottom border of each row
	else:
		print("No Board Created")

def intakeInteger():
	"""
	Only allows the input of integers
	Returns a "None" value otherwise
	"""
	intake = input()
	try:
		return int(intake)											# Tries to change input into an integer
	except ValueError:
		return None													# If failed, return None

def intakeOnly(vaildInputs):
	"""
	Only allows the input of values specified in the formal parameters
	Returns a "None" value otherwise
	"""
	intake = input()
	if intake in vaildInputs:										# Checks if input is in the list of verified inputs
		return intake

def intakecoords():
	"""
	Prompts the user for x and y coords of a position in the gameBoard
	"""
	print('\n')
	print(f'Where to place piece? (1-{len(gameBoard)},1-{len(gameBoard)})')
	while 1:
		print('X: ', end='')
		x = intakeInteger()	#	Gets an integer input
		if x != None and 1 <= x <= len(gameBoard):					#	Loop repeats if the inputted integer is out of the vaild range
			break
		print('Please enter a vaild number.')
	while 1:
		print('Y: ', end='')
		y = intakeInteger()	#	Gets an integer input
		if y != None and 1 <= y <= len(gameBoard):					#	Loop repeats if the inputted integer is out of the vaild range
			break
		print('Please enter a vaild number.')
	return (x,y)													#	Returns the intaken integers

def endCheck():
	"""
	Checks if any of the ending conditions for the game have been met
	"""
	Tie = True
	# Checks all possible horizonal line matches
	for y in range(len(gameBoard)):
		if gameBoard[y][0] != 2 and gameBoard[y].count(gameBoard[y][0]) == len(gameBoard): return gameBoard[y][0]	#	Checks if either X or O won horizonally
		elif Tie:
			for x in range(len(gameBoard)-1):
				if gameBoard[y][x] != 2 and ((gameBoard[y][x] + 1) % 2) in gameBoard[y][(x+1):]: break				#	Checks if either X or O can still win horizonally
			else: Tie = False
	
	# Checks all possible vertical line matches
	for x in range(len(gameBoard)):
		if gameBoard[0][x] != 2 and [row[x] for row in gameBoard].count(gameBoard[0][x]) == len(gameBoard): return gameBoard[0][x]	#	Checks if either X or O won vertically
		elif Tie:
			for y in range(len(gameBoard)-1):
				if gameBoard[y][x] != 2 and ((gameBoard[y][x] + 1) % 2) in [row[x] for row in gameBoard[(y+1):]]: break				#	Checks if either X or O can still win vertically
			else: Tie = False

	# Checks all possible diagonal line matches
	for i in [-1,0]:
		if gameBoard[i][0] != 2 and [gameBoard[z+(i*(2*z+1))][z] for z in range(len(gameBoard))].count(gameBoard[i][0]) == len(gameBoard): return gameBoard[i][0] 			#	Checks if either X or O won diagonally
		elif Tie:
			for z in range(len(gameBoard)-1):
				if gameBoard[z][z+(i*(2*z+1))] != 2 and ((gameBoard[z][z+(i*(2*z+1))] + 1) % 2) in [gameBoard[y][y+(i*(2*y+1))] for y in range(len(gameBoard)-z)]: break	#	Checks if either X or O can still win diagonally
			else: Tie = False

	if Tie:
		return 2	# If a tie occured, return 2

def minimax(coords, isMaximizing, piece, depth = 3, alpha = float('-inf'), beta = float('inf')):
	"""
	Uses the minimax algorithm, with alpha-beta pruning and a Transposition Table, to determine which future Boards are desirable
	"""
	from random import randint

	gameBoard[coords[1]][coords[0]] = (piece+isMaximizing)%2		#	Places a temporary piece on the board to visualize the current board situation
	result = endCheck()
	if result != None:												#	If current board repersents an end, return the board's end value
		gameBoard[coords[1]][coords[0]] = 2							#	Removes temporary piece, on the board, before function exit
		if result == 2: return 0
		elif result == piece: return 1
		elif result == ((piece+1)%2): return -1

	#	Acting player
	if isMaximizing:
		bestScore = -2	#	A placeholder bestScore that is less than the worst case
		for y in range(len(gameBoard)):
			for x in range(len(gameBoard)):
				if gameBoard[y][x] == 2:							#	Checks all positions that are vacant
					#	Finds the best path to an end for the Acting Player at this board position, while using slight randomness
					score = minimax((x,y), False, piece, depth-1, alpha, beta) + (randint(0, 15)/100)
					bestScore = max(score, bestScore)
					# Alpha Beta pruning
					alpha = max(score, alpha)
					if beta <= alpha: break
	#	Enemy Player
	else:
		bestScore = 2	#	A placeholder bestScore that is greater than the worst case
		for y in range(len(gameBoard)):
			for x in range(len(gameBoard)):
				if gameBoard[y][x] == 2:							#	Checks all positions that are vacant
					#	Finds the best path to an end for the Enemy Player at this board position, while using slight randomness
					score = minimax((x,y), True, piece, depth-1, alpha, beta) + (randint(-15, 0)/100)
					bestScore = min(score, bestScore)
					# Alpha Beta pruning
					beta = min(score, beta)
					if beta <= alpha: break

	gameBoard[coords[1]][coords[0]] = 2								#	Removes temporary piece, on the board, before function exit
	return bestScore

def AIBestMove(piece):
	"""
	Finds the best possible move, for the given piece, if it is the piece's current turn
	"""
	bestScore = -2		#	A placeholder bestScore that is less than the worst case
	bestMove = (-1,-1)	#	A placeholder invalid position
	
	for y in range(len(gameBoard)):
		for x in range(len(gameBoard)):
			if gameBoard[y][x] == 2:								#	Checks all positions that are vacant
				score = minimax((x,y), False, piece, 6)				#	Perform the minimax algorithm on these positions
				if bestScore < score:								#	Saves the best calculated moves from each minimax algorithm usage
					bestScore = score
					bestMove = (x+1,y+1)
	return bestMove


def placePiece(coords, piece):
	"""
	Places a single Tic Tac Toe piece on the board, if the given coordinate is available
	Return "True" if successful and "False" otherwise
	"""
	if all([1 <= coords[i] <= len(gameBoard) for i in range(len(coords))]) and gameBoard[coords[1]-1][coords[0]-1] == 2:
		gameBoard[coords[1]-1][coords[0]-1] = piece					#	Places a piece at position if position is on the board and vacant 
		return True
	else: return False												#	Else, return False

if __name__ == "__main__":
	import sys
	if len(sys.argv) == 1:											#	If there are only one given argument, default to a board length of three
		length = 3
	elif len(sys.argv) > 2:											#	If there are too many given arguments, quit out of the program
		print('Too many arguments.\nExpected format: python TicTacToe.py [integer between 3-5]')
		quit()
	else:
		try:
			length = int(sys.argv[1])
			if not (3 <= length <= 4 ):								#	If given argument is an integer outside of the vaild range, default to a board length of three
				print('Argument 1 out of valid range (3-4).\nDefaulting to 3...') 
				length = 3
		except ValueError:											#	If given argument is not an integer, quit out of the program
			print('Invalid argument 1.\nExpected format: python TicTacToe.py [integer between 3-4]') 
			quit()

	#	Loads up needed value references
	sprites = ( 'X', 'O', ' ' )
	endings = ( 'Player X has won.',
				'Player O has won.',
				'A tie has occured.'
			  )
	while 1:
		gameBoard = [[2 for _ in range(length)] for _ in range(length)]		#	Generates an enemy board
		currentTurn = False
		displayBoard()														#	Displays Board in ASCII
		while 1:
			#	Performs the current turn
			if not currentTurn:
				while not placePiece(AIBestMove(0), 0):
					print('\nThat position is unavailable. Try Again.')
			else:
				while not placePiece(AIBestMove(1), 1):
					print('\nThat position is unavailable. Try Again.')
			currentTurn = (not currentTurn)									#	Switches the current player
			displayBoard()													#	Displays Board in ASCII
			endType = endCheck()
			if endType != None:
				print(endings[endType])										#	Ends the game when applicable
				break
		print('Play again? (y/n)')
		intake = intakeOnly(['Y','y','N','n'])
		while intake not in ['Y','y']:										#	Restarts the game if either 'Y' or 'y' are entered
			if intake in ['N','n']: break									#	Exits the game if either 'N' or 'n' are entered
			print('Please enter a vaild input.\nPlay again? (y/n)')			#	Asks for input again if any other string is entered
			intake = intakeOnly(['Y','y','N','n'])
		else: continue
		break